pytorch_lightning/utilities/cli.py

```diff
@@ -591,11 +600,24 @@ def setup_parser(
 
     def add_default_arguments_to_parser(self, parser: LightningArgumentParser) -> None:
         """Adds default arguments to the parser."""
+
+        min_value = np.iinfo(np.int32).min
+        max_value = np.iinfo(np.int32).max
+        int32 = restricted_number_type("int32", int, [(">=", min_value), ("<=", max_value)])
```

Comment:
Don't define a type inside a method. This should be at the top of this file or in [pytorch_lightning/utilities/types.py](https://github.com/PyTorchLightning/pytorch-lightning/tree/master/pytorch_lightning/utilities/types.py). It needs to be inside an `if _JSONARGPARSE_AVAILABLE:` block since jsonargparse is an optional dependency.

After a second thought, we could do something even better. Instead of creating a new type, we use `np.int32`. But this type needs to be registered so that the parsing fails on invalid values.

```python
from jsonargparse.typing import register_type

int32_min = np.iinfo(np.int32).min
int32_max = np.iinfo(np.int32).max

def int32_deserializer(value):
    value = int(value)
    if value < int32_min or value > int32_max:
        raise ValueError(f"value outside int32 range: {value}")
    return value

register_type(np.int32, deserializer=int32_deserializer)
```

Reply:
The first paragraph is clear. I'm not sure if I understand the second paragraph correctly, as I'm not so familiar with the jsonargparse internals. 

Instead of defining the new type int32 we use the existing type np.int32. The deserializer converts the CLI/config/default value to the internal representation and throws an error if the value is out of range. What irritates me is that  there isn't a cast to np.int32... So we are just using the class_name np.int32 but actually we are just using ints?

So if I understand this correctly, we will then set the type of `seed_everything_default` to `Optional[Union[np.int32, bool]]` instead of `Optional[Union[int/int32, bool]]` and the type of `--seed_everything` to `Union[bool, np.int32] `instead of` Union[bool, int/int32]` ? But in this case the linter will throw warnings if the user passes an int as argument to seed_everything_default.... 

For now I set it to uint32.

Reply:
If `seed_everything_default` has as type `Optional[Union[np.int32, bool]]` and someone wants to give a value, they will need to cast it to not get a warning, e.g. `seed_everything_default=np.int32(9876)`. Either this or add a comment to disable the warning.

Reply:
But is forcing the user to cast their `int `seed to `numpy.uint32` (in a pytorch based library) the right way? 
I think the current implementation, with the custom type, is the more user friendly implementation.

If you really think the implementation with the custom type shouldn't be used, an alternative would be to set the argument type of `seed_everything_default` back to `int `(like it was before.. the `seed_everything` function also takes an `int`, so I dont think these two arguments should have different types) and enforce the correct range just in the ArgumentParser with your `register_type(np.int32, deserializer=int32_deserializer)` line.

Reply:
This is just how type hints work in python. The value you give must be one of the accepted types. If `int` is not an accepted type, then the value must be casted. It is the same if it is a type created with `restricted_number_type` or `np.int32`, both require to cast. The only benefit of `np.int32` is that a well known type is used instead of creating an equivalent one for this single use case.

The only reason that I mentioned all of this is because you said in #12804

> This change also has the advantage that it prevents passing an int greater than pytorchs int32.max, which results, without raising a warning, in using a different random seed (same behaviour for numpy).

Can you please clarify the problem exactly. Did I understand correctly that giving to `seed_everything` a value `>int32.max` will give different results each time it is run? i.e. it becomes non-deterministic. If this is the case, it does seem like a problem to fix. And not just to fix in the cli, but also in the `seed_everything` function. Also I don't understand why you are now using `uint32`. `np.iinfo(np.uint32).max` is not the same as `np.iinfo(np.int32).max` and with `uint32` no negative is allowed. Which range of values are the ones that create problems?

If for any int value the runs are deterministic, even `>int32.max`, then the type should be kept as `int`.

Reply:
The switch from `in32` to `uint32` is because I made a mistake in the original issue. It should have been uint32.

Regarding passing in a wrong seed, if a a value `>np.uint32.max` is passed in the behaviour will be non-deterministic and the pytorch and numpy seeding functions in this case do not throw an warning.
However, I just looked in `seed_everything()` and it seems it already throws a warning (there could be a discussion if it should raise an error instead, but thats not the point of this PR), which I seemed to have missed in the output, when I had an issue with out of range seeds. In this case I would now change back the type of `seed_everything_default` and `--seed_everything` to `int`, as warning would already be raised by `seed_everything()`. Sorry for the confusion.